pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -1440,28 +1440,40 @@ func validateFirmwareACPI(field *k8sfield.Path, spec *v1.VirtualMachineInstanceS
 	}
 
 	acpi := spec.Domain.Firmware.ACPI
-	for _, volume := range spec.Volumes {
-		if acpi.SlicNameRef != volume.Name {
+	causes = append(causes, validateACPIRef(field, acpi.SlicNameRef, spec.Volumes, "SlicNameRef")...)
+	causes = append(causes, validateACPIRef(field, acpi.MsdmNameRef, spec.Volumes, "MsdmNameRef")...)
+	return causes
+}
+
+func validateACPIRef(field *k8sfield.Path, nameRef string, volumes []v1.Volume, fieldName string) []metav1.StatusCause {
+	var causes []metav1.StatusCause
+	if nameRef == "" {
+		return causes
+	}
+
+	found := false
+	for _, volume := range volumes {
+		if nameRef != volume.Name {
 			continue
 		}
-
-		switch {
-		case volume.Secret != nil:
-		default:
+		found = true
+		if volume.Secret == nil {
 			causes = append(causes, metav1.StatusCause{
 				Type:    metav1.CauseTypeFieldValueInvalid,
 				Message: fmt.Sprintf("%s refers to Volume of unsupported type.", field.String()),
-				Field:   field.Child("slicNameRef").String(),
+				Field:   field.Child(nameRef).String(),
 			})
 		}
-		return causes
+		break
 	}
 
-	causes = append(causes, metav1.StatusCause{
-		Type:    metav1.CauseTypeFieldValueInvalid,
-		Message: fmt.Sprintf("%s does not have a matching Volume.", field.String()),
-		Field:   field.String(),
-	})
+	if !found {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s does not have a matching Volume.", field.String()),
+			Field:   field.Child(fieldName).String(),
+		})
+	}
```

Comment:
An alternative:
```golang
func validateACPIRef(field *k8sfield.Path, nameRef string, volumes []v1.Volume, fieldName string) []metav1.StatusCause {
	if nameRef == "" {
		return nil
	}

	for _, volume := range volumes {
		if nameRef != volume.Name {
			continue
		}

		if volume.Secret != nil {
			// all good
			return nil
		}

		return []metav1.StatusCause{{
			Type:    metav1.CauseTypeFieldValueInvalid,
			Message: fmt.Sprintf("%s refers to Volume of unsupported type.", field.String()),
			Field:   field.Child(fieldName).String(),
		}}
	}

	return []metav1.StatusCause{{
		Type:    metav1.CauseTypeFieldValueInvalid,
		Message: fmt.Sprintf("%s does not have a matching Volume.", field.String()),
		Field:   field.Child(fieldName).String(),
	}}
}
```

Reply:
In this case, if nameRef matches it means that this is volume to be used by either slic or msdm. It *must* be a secret, so returning early is no good.

Reply:
scratch that, I misread. If volume.Secret is set, we are all good indeed.